engine/src/main/java/org/pentaho/di/job/entries/ftp/JobEntryFTP.java

```diff
@@ -0,0 +1,1528 @@
+      ftpclient.setRemoteAddr( getInetAddress( realServername ) );
+      if ( !Utils.isEmpty( realServerPort ) ) {
+        ftpclient.setRemotePort( Const.toInt( realServerPort, 21 ) );
+      }
+
+      if ( !Utils.isEmpty( proxyHost ) ) {
+        String realProxy_host = environmentSubstitute( proxyHost );
+        ftpclient.setRemoteAddr( InetAddress.getByName( realProxy_host ) );
+        if ( isDetailed() ) {
+          logDetailed( BaseMessages.getString( PKG, "JobEntryFTP.OpenedProxyConnectionOn", realProxy_host ) );
+        }
+
+        // FIXME: Proper default port for proxy
+        int port = Const.toInt( environmentSubstitute( proxyPort ), 21 );
+        if ( port != 0 ) {
+          ftpclient.setRemotePort( port );
+        }
+      } else {
+        ftpclient.setRemoteAddr( getInetAddress( realServername ) );
+
+        if ( isDetailed() ) {
+          logDetailed( BaseMessages.getString( PKG, "JobEntryFTP.OpenedConnectionTo", realServername ) );
+        }
+      }
+
+      // set activeConnection connectmode ...
+      if ( activeConnection ) {
+        ftpclient.setConnectMode( FTPConnectMode.ACTIVE );
+        if ( isDetailed() ) {
+          logDetailed( BaseMessages.getString( PKG, "JobEntryFTP.SetActive" ) );
+        }
+      } else {
+        ftpclient.setConnectMode( FTPConnectMode.PASV );
+        if ( isDetailed() ) {
+          logDetailed( BaseMessages.getString( PKG, "JobEntryFTP.SetPassive" ) );
+        }
+      }
+
+      // Set the timeout
+      ftpclient.setTimeout( timeout );
+      if ( isDetailed() ) {
+        logDetailed( BaseMessages.getString( PKG, "JobEntryFTP.SetTimeout", String.valueOf( timeout ) ) );
+      }
+
+      ftpclient.setControlEncoding( controlEncoding );
+      if ( isDetailed() ) {
+        logDetailed( BaseMessages.getString( PKG, "JobEntryFTP.SetEncoding", controlEncoding ) );
+      }
+
+      // If socks proxy server was provided
+      if ( !Utils.isEmpty( socksProxyHost ) ) {
+        if ( !Utils.isEmpty( socksProxyPort ) ) {
+          FTPClient.initSOCKS( environmentSubstitute( socksProxyPort ), environmentSubstitute( socksProxyHost ) );
+        } else {
+          throw new FTPException( BaseMessages.getString(
+            PKG, "JobEntryFTP.SocksProxy.PortMissingException", environmentSubstitute( socksProxyHost ),
+            getName() ) );
+        }
+        // then if we have authentication information
+        if ( !Utils.isEmpty( socksProxyUsername ) && !Utils.isEmpty( socksProxyPassword ) ) {
+          FTPClient.initSOCKSAuthentication(
+            environmentSubstitute( socksProxyUsername ), Utils.resolvePassword( this, socksProxyPassword ) );
+        } else if ( !Utils.isEmpty( socksProxyUsername )
+          && Utils.isEmpty( socksProxyPassword ) || Utils.isEmpty( socksProxyUsername )
+          && !Utils.isEmpty( socksProxyPassword ) ) {
+          // we have a username without a password or vica versa
+          throw new FTPException( BaseMessages.getString(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Exception [FTPException] is explicitly thrown on this line but is caught in the outer catch clause with generic exception on line [1153].  Consider adding a catch clause of this specific exception in the outer try statement.